### PR TITLE
[autoscaler][kuberay] Disable autoscaler health check and drain functionality 

### LINKED ISSUE #26379: [KubeRay][Autoscaler] Provide a flag to disable worker health-check.
### What happened + What you expected to happen

The Ray autoscaler health-checks Ray nodes by polling the GCS for Raylet resource data -- if the autoscaler doesn't get resources for long enough, it kills the Ray node.

This behavior is redundant with the health-check that will soon be implemented in KubeRay.
The task is to provide a flag that will disable this behavior when autoscaling with KubeRay.
Healthchecks should be left to KubeRay + Kubernetes.

### Versions / Dependencies

N/A

### Reproduction script

N/A

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

In the context of KubeRay deployments, the autoscaler doesn't have any business managing worker lifecycle, beyond scale-up and scale-down.

This PR toggles off two autoscaler functionalities that are better left to the KubeRay operator (especially given that the autoscaler is an optional component in this scheme.)
- **Termination of Ray nodes disconnected from the GCS**. The autoscaler tracks the last time it has gotten a resource message for a node from the GCS and terminates the worker if the last message came more than 30 seconds ago. Worker health checks probes are best left to the operator and K8s, as in https://github.com/ray-project/kuberay/pull/294. Moreover, in the context of GCS HA, it is incorrect for the autoscaler to downscale a Ray node disconnected from the GCS.
- **Ray node drain by RPC**. Prior to downscaling a Ray node, the autoscaler asks the GCS to "drain" a Ray node, where the current implementation of "drain" is actually "kill". This kind of pre-scale-down preparation is better handled by a `ray stop` [preStop hook](https://github.com/ray-project/kuberay/blob/e1e41c30e6e74f279bdfa8de2e1cc37298d5f91f/ray-operator/config/samples/ray-cluster.complete.yaml#L161-L163) which is what is currently done in KubeRay. Trying to kill the Ray node by RPC from the autoscaler may lead to weird interactions where the autoscaler is trying to kill the Ray node while K8s is trying to restart the Ray container.

The PR introduces two flags and sets them to `False` for the KubeRay NodeProvider.

## Related issue number

Closes https://github.com/ray-project/ray/issues/26379

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_autoscaler.py [hunk 1]
```diff
@@ -1567,7 +1578,7 @@ def helperDynamicScaling(
             DrainNodeOutcome.GenericException,
         ):
             # DrainNode call was made.
-            mock_node_info_stub.drain_node_call_count > 0
+            assert mock_node_info_stub.drain_node_call_count > 0
```
Comment (DmitriGekhtman) [inline python/ray/tests/test_autoscaler.py:1581]: We should add some linters to catch useless comparisons.

**Comment by brucez-anyscale** [comment]
I am not familiar with autoscaler code. Will let others to review.

**Comment by DmitriGekhtman** [comment]
TODO: Make sure with these changes there are no weird error messages when a node is scaled down. IF weird error messages appear, just keep the drain RPC for now, then implement a better pre-stop hook, and turn off the "drain node RPC" then.

**Comment by pcmoritz** [comment]
@brucez-anyscale Can you review the health check portion in the context of https://github.com/ray-project/kuberay/pull/294? Even though you are not familiar with the autoscaler code, it would be helpful to get your take here / make sure it makes sense in the light of the other PR.

File: python/ray/tests/test_autoscaler.py [hunk 2]
```diff
@@ -2717,24 +2765,40 @@ def testTerminateUnhealthyWorkers(self):
...
-        events = autoscaler.event_summarizer.summary()
-        assert (
-            "Removing 1 nodes of type "
-            "ray-legacy-worker-node-type (lost contact with raylet)." in events
-        ), events
+        if disable_liveness_check:
+            # We've disabled the liveness check, so the unhealthy node should stick
+            # around until someone else takes care of it.
+            # Do several autoscaler updates:
+            for _ in range(10):
```
Comment (pcmoritz) [inline python/ray/tests/test_autoscaler.py:2773]: I don't understand why there is a loop here

File: python/ray/tests/test_autoscaler.py [hunk 3]
```diff
@@ -2717,24 +2765,40 @@ def testTerminateUnhealthyWorkers(self):
...
+                autoscaler.update()
+            # The nodes are still there.
+            assert self.num_nodes() == 2
```
Comment (pcmoritz) [inline python/ray/tests/test_autoscaler.py:2776]: Do we need some sort of synchronization before this is effective?

File: python/ray/tests/test_autoscaler.py [hunk 4]
```diff
@@ -2717,24 +2765,40 @@ def testTerminateUnhealthyWorkers(self):
...
+            # Check the node removal event is generated.
+            autoscaler.update()
+            events = autoscaler.event_summarizer.summary()
+            assert (
+                "Removing 1 nodes of type "
+                "ray-legacy-worker-node-type (lost contact with raylet)." in events
+            ), events
+
+            # No additional runner calls, since updaters were disabled.
+            time.sleep(1)
```
Comment (pcmoritz) [inline python/ray/tests/test_autoscaler.py:2798]: Uhm... I know this was here before but it has some pretty bad optics :(

**Review by pcmoritz — APPROVED** [review]

File: python/ray/tests/test_autoscaler.py (hunk 4 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler.py:2798]: I will take a look into including the optics on this one.
TBH, this entire file is pretty bad optics.

File: python/ray/tests/test_autoscaler.py (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler.py:2773]: Strictly speaking, it's not necessary. 

I'll add a comment to this effect:
We're demonstrating that the autoscaler will never take action to remove the unhealthy node, no matter how long it runs.

File: python/ray/tests/test_autoscaler.py (hunk 3 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler.py:2776]: I will add a comment explaining this: Node deletion is synchronous and blocking, as are the methods of the mock node provider used in this test.

The sort of synchronization that would be nice is to make the autoscaler single-threaded in all contexts...

**Review by wuisawesome — CHANGES_REQUESTED** [review]
Reminder to make sure that removing the drain node call doesn't lead to an unexpected node failure to be reported pls

**Comment by DmitriGekhtman** [comment]
@wuisawesome It is perhaps surprising, but it appears the `ray stop` preStop hook is working as expected -- there is no warning message.

I ran an image built out of this PR in a Ray cluster consisting of 2 nodes with 1 CPU each and the capacity to add an extra 1 CPU node. I set the idle timeout to seconds.
I ran the following script and saw no odd error messages in the driver logs (just the expect scale up and scale down messages).
```
import ray
import time

ray.init()

@ray.remote(num_cpus=1)
class Clog:
    def __init__(self):
        pass

@ray.remote(num_cpus=1)
def f():
    return "hi"

# Clog the available worker with 1-CPU actors.
actor_ref1 = Clog.remote()
actor_ref2 = Clog.remote()

# Force scale-up and immediate scale-down with a short-lived task.
print(ray.get(f.remote()))
# Wait for concerning driver log spew.
time.sleep(1000)
# ...looks like we're good
```

File: python/ray/tests/test_autoscaler.py (hunk 4 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler.py]: I fixed the optics by removing the time.sleep and adding an autoscaler subclass that raises an assertion error if you try to create node updaters with it.

File: python/ray/tests/test_autoscaler.py (hunk 3 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler.py:2789]: Added comments about why no synchronization is required.

File: python/ray/tests/test_autoscaler.py (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler.py:2786]: Added the comment.

**Review by wuisawesome — APPROVED** [review]

**Comment by DmitriGekhtman** [comment]
> Reminder to make sure that removing the drain node call doesn't lead to an unexpected node failure to be reported pls

Alex checked to confirm that the Raylet's sigterm handler interacts correctly with the GCS and avoids this issue. Will merge assuming no test issues.

**Pull request merged**